### PR TITLE
Task 2.3 - class Item

### LINKED ISSUE #42: Task 2.3 - class Item (methods)
W klasie Item dopisz brakujące metody oraz ich implementacje.

```
// override from Cargo
size_t getPrice() const override;
std::string getName() const override { return name_; }
size_t getAmount() const override { return amount_; }
size_t getBasePrice() const override { return base_price_; }
Cargo& operator+=(size_t amount) override;
Cargo& operator-=(size_t amount) override;
bool operator==(Cargo& cargo) const override;
```

### PR DESCRIPTION
**Pull request by mariuszlisowski**
Driver code included for testing purposes (in test directory)

### CODE REVIEW

File: shm/src/item.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#include "item.hpp"
```
Comment (m-kop) [inline shm/src/item.cpp:1]: please add this file to CMakeLists.txt

File: shm/src/item.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+std::string Item::getName() const { return name_; }
+size_t Item::getAmount() const { return amount_; }
+size_t Item::getBasePrice() const { return basePrice_; }
+Item::Rarity Item::getRarity() const { return rairty_; }
+
+bool Item::operator==(const Cargo& cargo) const {
+    if (typeid(cargo) == typeid(Item)) {
+        const Item* rarityPtr = static_cast<const Item*>(&cargo);
```
Comment (m-kop) [inline shm/src/item.cpp]: the variable name is misleading

Comment (m-kop) [reply shm/src/item.cpp]: rarityPtr?

File: shm/src/item.cpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        return name_ == rarityPtr->getName() &&
+               amount_ == rarityPtr->getAmount() &&
+               basePrice_ == rarityPtr->getBasePrice() &&
+               rairty_ == rarityPtr->getRarity();
+    }
+    return false;
+}
+bool Item::operator!=(const Cargo& cargo) const {
+    if (typeid(cargo) == typeid(Item)) {
+        const Item* rarityPtr = static_cast<const Item*>(&cargo);
```
Comment (m-kop) [inline shm/src/item.cpp]: ditto

**Review by m-kop — CHANGES_REQUESTED** [review]
good work!

File: shm/inc/item.hpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+class Item : public Cargo {
+public:
+    enum class Rarity {
+        common = 1,
+        rare = 4,
+        epic = 8,
+        legendary = 16
+    };
+
+    Item(const std::string name, size_t amount, size_t basePrice, Rarity rarity)
```
Comment (m-kop) [inline shm/inc/item.hpp]: const & for string

File: shm/src/item.cpp (hunk 1 above)
Comment (mariuszlisowski) [reply shm/src/item.cpp:1]: Ok

File: shm/inc/item.hpp (hunk 4 above)
Comment (mariuszlisowski) [reply shm/inc/item.hpp]: Right, done.

File: shm/src/item.cpp (hunk 2 above)
Comment (mariuszlisowski) [reply shm/src/item.cpp]: Yes, it is misleading. I can see now. I must had been thinking about the enum class 'rarity' at the moment and left it unchanged. Not it is simply 'item'. Or maybe ptr_item would be better? Let me know if so.

**Pull request merged**